src/renderers/dom/fiber/ReactDOMFiberComponent.js

```diff
@@ -69,6 +70,49 @@ if (__DEV__) {
     validateInputPropertes(type, props);
     validateUnknownPropertes(type, props);
   };
+
+  var warnForTextDifference = function(serverText: string, clientText: string) {
+    if (didWarnInvalidHydration) {
+      return;
+    }
+    didWarnInvalidHydration = true;
+    warning(
+      false,
+      'Text content did not match. Server: "%s" Client: "%s"',
+      serverText,
+      clientText,
+    );
+  };
```

Comment:
Even if hoisted, this assignment would get stripped, so it'd still get DCE.

We have had a few warnOnce solutions but the context of "once" varies a lot so hasn't been worth the abstraction cost. A few:

- Once per React module instance.
- Once per component class.
- Once per component instance.
- Once per subtree.

These can then combine with:

- Once per each type of warning.
- Once for any of multiple types of warnings.
- Etc.

In this case, we should ideally gather all of these warnings up into one actionable "diff view" of all the differences. That's on my todo. To avoid spam, right now I just want to show the first diff among any type of diff.